Modify container handling for GTK

freakboy3742 opened the pull request:
In the process of performing the audit of Button, it became clear that GTK had some fundamental issues with the layout of widgets, and the root widget in particular. 

Firstly, the Pack layout was somewhat confused around how it was meant to behave with the "root" content of a layout. Using HTML as a model, the "root" content is the `<body>` element; and that element is always the size of the display window. Previously, Pack would only assume the full size of the window if the root content had no children; as soon as the root content had children, the flex axis (i.e., vertical in a Column box, horizontal in a Row box) would be shrunk to the minimum possible size. This led to interesting situations where a Column layout would be flexible in height, but wouldn't expand to the full height of the window. It would also lead to background colors *not* being applied over the entire background.

This PR makes an attempt at a policy statement at the intended Pack->HTML/CSS mapping as a source of reference, and then implements some changes to the handling of the root element that reflects those changes.

This PR also performs a radical change to how layout is achieved on GTK; however, this change is much simpler to implement, and fixes a lot of edge case bugs.

 * A new TogaContainer object has been introduced. This is a GTK-specific object that is a GTK widget implementing the Pack layout.
 * The Window has an internal Box layout that is used to store the toolbar, and the TogaContainer.
 * Instead of the root widget of Toga layout also serving as the native container, all Toga content is added to the TogaContainer.
 * The TogaContainer merges 2 historical concepts: Container and Viewport. 
 * Any change in size or exposure of the TogaContainer triggers a GTK layout. This is used to re-compute the minimum and preferred layout sizes, and to re-layout the container.
 * Rehinting can't be reliably performed until a the TogaContainer (the widget implementing layout) is ready to be laid out. As a result, the Toga-level `rehint()` request doesn't do an immediate layout; instead, it adds the widget to a list of "dirty" widgets. When a layout occurs, all dirty widgets are *actually* re-hinted, ensuring that the hints are the right size going into a layout.
 * Changing a style property (like `flex` or `width`) now triggers a re-layout. This means there should be no user-space need ever to invoke `refresh()`.
 * The TogaContainer tracks whether it is dirty and needs a redraw; this property can be used by testing tools to evaluate whether all changes in style have been reflected in the layout.
 
We should also investigate whether this change (or platform-appropriate variations) should be made on other backends. I strongly suspect that the root widget handling is subtly wrong on most platforms (except GTK, at this point). As an interim measure, the implementation maintains an interface that allows for `_impl.container` and `_impl.viewport`, even though they are now the same object on the GTK backend.

Fixes #1205.

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

mhsmith commented:
> Using HTML as a model, the "root" content is the `<body>` element; and that element is always the size of the display window.

Based on my experiments, this is only the case: when quirks mode is active, which happens when the document contains no DOCTYPE, or a very old one. I think this is what was happening with your test document on Thursday. 

With `<!DOCTYPE html>`, the `<html>` and `<body>` elements follow the normal rules for block elements, and shrink vertically to fit their content. However, it may not look that way, because there's a special case in the CSS spec which says the `background-color` and `overflow` properties on the `<body>` element are actually applied to the viewport and not the body at all. If you want to see what the `<body>` dimensions really are, you can add a border, or use the browser's layout inspector.

So it looks like we should add the following rule:
```
html, body {
    height: 100%;
}
```

I'll post the rest of my review tomorrow.

freakboy3742 commented:
Agreed that quirks mode was the problem here; the `100%` rule change is definitely required. I'll make a change to that effect.

freakboy3742 commented:
I've done some more fiddling with quirks mode, and updated the mapping definition to include a full reference HTML document.

I've also added `white-space: nowrap` to the body definition, on the basis that we don't do content wrapping anywhere in Pack.

Based on those changes, Tutorial 0 becomes:
```
<body style="flex-direction: row;">
    <div style="flex-direction: row; margin: 50px; flex: 1 0 0%">Press me</div>
</body>
```

and Tutorial 1 becomes:
<body style="flex-direction: column; margin: 20px">
    <div style="flex-direction: row; margin: 10px">
        <div style="margin-left: 210px; flex:1 0 0%;">Input</div>
        <div style="margin-left: 10px; width: 200px;">Celcius</div>
    </div>
    <div style="flex-direction: row; margin: 10px">
        <div style="width: 200px; margin-right: 10px">Is equal to</div>
        <div style="flex:1 0 0%;">Input</div>
        <div style="margin-left: 10px; width: 200px">Fahrenheit</div>
    </div>
    <div style="margin: 15px">Convert</div>
</body>
```
(This tutorial 1 example has slightly different margin values, but the effect is the same).

Interestingly, I think I've found a bug in Chrome (109.0.5414.119) around the handling of margin collapse on the root element. Opening the document in a new tab always gives the correct rendering; however, `flex-direction:column` containers collapse the root element's top margin if the document is already loaded on the page.

mhsmith commented on docs/reference/style/pack.rst:
```diff
@@ -465,3 +458,104 @@ specifying the allocated width and allocated height.
+                  display: flex;
+                  margin: 0;
+               }
+               div {
+                  display: inline-flex;
+                  margin: 0;
```
```suggestion
               div {
                  display: flex;
```
According to [this](https://www.w3.org/TR/css-flexbox-1/#flex-items), "The display value of a flex item is blockified", so there's no difference between `flex` and `inline-flex` in this context.

There's also no need for a `margin`, because unlike `body`, no browser will have a default margin on `div`.

freakboy3742 replied:
Agreed on `margin`; I swear there was a reason I used inline-flex, but that appears to be lost to the depths of time. I can't find any case where it matters.

mhsmith commented on docs/reference/style/pack.rst:
```diff
@@ -465,3 +458,104 @@ specifying the allocated width and allocated height.
+      <!DOCTYPE html>
+      <html>
+         <head>
+           <meta charset="UTF-8" />
+            <title>Pack layout testbed</title>
```
```suggestion
            <meta charset="UTF-8" />
            <title>Pack layout testbed</title>
```
Indentation.

mhsmith commented on docs/reference/style/pack.rst:
```diff
@@ -465,3 +458,104 @@ specifying the allocated width and allocated height.
+                                 otherwise ignored.
+   ``alignment: center``         ``align-items: center``
+   ``direction: <str>``          ``flex-direction: <str>``
+   ``display: pack``             ``display: flex`` on the root element of a layout;
+                                 ``display: inline-flex`` on all other elements.
```
```suggestion
   ``display: pack``             ``display: flex``
```
See above.

mhsmith commented on docs/reference/style/pack.rst:
```diff
@@ -465,3 +458,104 @@ specifying the allocated width and allocated height.
+                                 ``display: inline-flex`` on all other elements.
+   ``flex: <int>``               If ``direction = row`` and ``width`` is set,
+                                 or ``direction = column`` and ``height`` is set,
+                                 ignore. Otherwise, ``flex: <int> 0 0%``.
```
I think it's confusing to use a % sign here, because it would be uncommon to use a percentage in this context. The examples in the standard [here](https://www.w3.org/TR/css-flexbox-1/#valdef-flex-flex-basis) and [here](https://www.w3.org/TR/css-flexbox-1/#flex-initial) simply say "0".

This also applies to the other % signs below, and in the `__css__` method in pack.py.

freakboy3742 replied:
Weird - I thought I saw something on MDN that rendered the default flex as `0 0 0%`... but now I can't find it. Updated here, and in `__css__` to use `0`.

freakboy3742 commented on core/src/toga/style/pack.py:
```diff
@@ -526,6 +532,56 @@ def _layout_column_children(self, node, available_width, available_height, scale
+            css.append("flex: 0 0 0%;")
+
+        # flex
+        if self.flex:
+            css.append(f"flex: {self.flex} 0 0%;")
```
I believe I was relying on the fact that a flex definition will be ignored if width/height are defined. However, to avoid confusion, it's probably better to merge this into the width/height constructions above.

mhsmith replied:
Based on my tests, both `flex: 1 0 0` and `flex: 0 0 0` do have an effect in CSS even if there's an explicit width/height, so we'd better only generate one of them.

freakboy3742 commented on core/tests/style/test_pack.py:
```diff
@@ -314,45 +320,16 @@ def test_tutorial_1(self):
-                ],
-            },
-        )
+        # # Minimum size÷å
```
Not sure what happened here, but yes - this is a mistake.

freakboy3742 commented on docs/reference/style/pack.rst:
```diff
@@ -465,3 +458,104 @@ specifying the allocated width and allocated height.
+            <style>
+               html, body {
+                  height: 100%;
+                  white-space: nowrap;
```
Agreed that `pre` is a better choice; however, it appears we can't apply it here, because it alters the rendering of `<html>` for some reason.

mhsmith commented on core/src/toga/style/pack.py:
```diff
@@ -565,21 +565,17 @@ def __css__(self):
+        # width/flex
         if self.width:
             css.append(f"width: {self.width}px;")
         elif self.direction == ROW and not self.flex:
```
```suggestion
        elif self.direction == ROW:
```

mhsmith commented on docs/reference/style/pack.rst:
```diff
@@ -238,230 +238,90 @@ The weight of the font to be used.
+   ``width: <int>``              ``width: <int>px``
+   ============================= ===================================================
+
+* If ``direction = row``, and neither ``width`` nor ``flex`` is set, the element
+  assumes a CSS definition of ``flex: 0 0 0``.
+
+* If ``direction = column``, and neither ``height`` nor ``flex`` is set, the
+  element assumes a CSS definition of ``flex: 0 0 0``.
+
+* All other Pack declarations should be used as-is as CSS declarations, with
```
```suggestion
* All other Pack declarations should be used as-is as CSS declarations, with
```
Since the initial value of `flex` is 0, this behavior is already covered by the `flex` definition above, as the `__css__` implementation shows. So these lines are redundant. This is consistent with the way we defined other properties with initial values, such as `alignment`.

mhsmith commented:
I think there will be some overlap between the GTK changes and what I'm working on for Android. So I've reduced this PR to the CSS mapping changes, which I'll merge now, and I'll make a separate PR for the GTK changes.
